File: backend/app/adapters/crm/hubspot.py
"""
HubSpot API client implementing CrmClient (Python).

Uses httpx async.

Mirrors src/crm/hubspot.ts EXACTLY:
- v3 objects POST /crm/v3/objects/{contacts,deals,tasks}
- associations arrays (HUBSPOT_DEFINED + typeId 3 for deal-contact, 216 for task-deal)
- hs_timestamp = ms epoch UTC at 09:00Z
- properties: firstname/lastname/jobtitle/company/email/hs_linkedin_url , dealname/description/pipeline/dealstage , hs_task_*
- name split, error msgs, {id: str} returns

See ARCHITECTURE.md, contracts/crm-client.md, crm_client.py for details.
Constructor takes access token (Bearer).
"""

from datetime import datetime
from typing import Any

import httpx

from .types import CrmClient, CrmContactInput, CrmDealInput, CrmTaskInput, CrmId


class HubspotClient:
    """Implements CrmClient Protocol for HubSpot."""

    BASE_URL = "https://api.hubapi.com"

    def __init__(self, access_token: str):
        if not access_token or not isinstance(access_token, str):
            raise ValueError("HubSpot access token is required")
        self.access_token = access_token

    # ─── Core HTTP ───────────────────────────────────────────────────────────────

    async def _call(self, path: str, body: dict[str, Any]) -> dict[str, Any]:
        url = f"{self.BASE_URL}{path}"
        headers = {
            "Authorization": f"Bearer {self.access_token}",
            "Content-Type": "application/json",
        }

        async with httpx.AsyncClient(timeout=30.0) as client:
            try:
                response = await client.post(url, json=body, headers=headers)
            except Exception as err:
                raise RuntimeError(f"Network error calling HubSpot ({path}): {err}") from err

            try:
                body_json: dict[str, Any] = response.json()
            except Exception:
                body_json = {}

            if response.status_code >= 400:
                msg = body_json.get("message") or body_json.get("error") or response.reason_phrase
                corr = f" (corr: {body_json.get('correlationId')})" if body_json.get("correlationId") else ""
                raise RuntimeError(f"HubSpot HTTP {response.status_code} on {path}: {msg}{corr}")

            if body_json and body_json.get("status") == "error":
                raise RuntimeError(
                    f"HubSpot API error on {path}: {body_json.get('message', 'unknown error')}"
                )

            if "id" not in body_json:
                raise RuntimeError(f"HubSpot returned no id for {path}")

            return body_json

    # ─── Helpers ─────────────────────────────────────────────────────────────────

    def _split_name(self, full: str) -> tuple[str, str]:
        """Split name like Bitrix24Client (first + rest as last)."""
        parts = full.strip().split()
        return (
            parts[0] if parts else full,
            " ".join(parts[1:]) if len(parts) > 1 else "",
        )

    def _to_hubspot_timestamp(self, due_date: str) -> int:
        """Convert YYYY-MM-DD or ISO to ms-since-epoch UTC at 09:00Z (parity with Bitrix)."""
        base = due_date.split("T")[0] if "T" in due_date else due_date
        iso = f"{base}T09:00:00.000Z"
        try:
            # datetime.fromisoformat needs care for Z; use timestamp
            dt = datetime.fromisoformat(iso.replace("Z", "+00:00"))
            return int(dt.timestamp() * 1000)
        except Exception:
            return int(datetime.utcnow().timestamp() * 1000)

    # ─── Contact ─────────────────────────────────────────────────────────────────

    async def createContact(self, contact: CrmContactInput) -> CrmId:
        if isinstance(contact, dict):
            name = contact.get("name", "")
            role = contact.get("role")
            company = contact.get("company", "")
            email = contact.get("email")
            linkedin = contact.get("linkedin")
        else:
            name = contact.name
            role = contact.role
            company = contact.company
            email = contact.email
            linkedin = contact.linkedin

        firstname, lastname = self._split_name(name)

        properties: dict[str, str] = {
            "firstname": firstname,
            "lastname": lastname,
            "company": company,
        }
        if role:
            properties["jobtitle"] = role
        if email:
            properties["email"] = email
        if linkedin:
            properties["hs_linkedin_url"] = linkedin

        result = await self._call(
            "/crm/v3/objects/contacts",
            {"properties": {k: v for k, v in properties.items() if v}},
        )

        return {"id": result["id"]}

    # ─── Deal ─────────────────────────────────────────────────────────────────────

    async def createDeal(self, deal: CrmDealInput) -> CrmId:
        if isinstance(deal, dict):
            title = deal["title"]
            contact_id = deal["contactId"]
            comments = deal["comments"]
        else:
            title = deal.title
            contact_id = deal.contactId
            comments = deal.comments

        properties: dict[str, str] = {
            "dealname": title,
            "description": comments,
            "pipeline": "default",
            "dealstage": "appointmentscheduled",
        }

        body = {
            "properties": properties,
            "associations": [
                {
                    "to": {"id": contact_id},
                    "types": [
                        {
                            "associationCategory": "HUBSPOT_DEFINED",
                            "associationTypeId": 3,
                        }
                    ],
                }
            ],
        }

        result = await self._call("/crm/v3/objects/deals", body)

        return {"id": result["id"]}

    # ─── Task ─────────────────────────────────────────────────────────────────────

    async def createTask(self, task: CrmTaskInput) -> CrmId:
        if isinstance(task, dict):
            title = task["title"]
            description = task["description"]
            due_date = task["dueDate"]
            deal_id = task["dealId"]
        else:
            title = task.title
            description = task.description
            due_date = task.dueDate
            deal_id = task.dealId

        hs_timestamp = self._to_hubspot_timestamp(due_date)

        properties: dict[str, Any] = {
            "hs_task_subject": title,
            "hs_task_body": description,
            "hs_timestamp": hs_timestamp,
            "hs_task_status": "NOT_STARTED",
        }

        associations = [
            {
                "to": {"id": deal_id},
                "types": [
                    {
                        "associationCategory": "HUBSPOT_DEFINED",
                        "associationTypeId": 216,
                    }
                ],
            }
        ]

        result = await self._call(
            "/crm/v3/objects/tasks",
            {"properties": properties, "associations": associations},
        )

        return {"id": result["id"]}

File: src/crm/hubspot.ts
/**
 * HubSpot API client implementing CrmClient
 * Uses Node.js 18+ native fetch — no axios, no node-fetch.
 *
 * HubSpot v3 REST specifics (per ARCHITECTURE.md, contracts/crm-client.md, research):
 * - Base: https://api.hubapi.com
 * - Auth: Authorization: Bearer <accessToken> (OAuth or Private App token)
 * - Objects: /crm/v3/objects/contacts , /deals , /tasks
 * - Create with associations (for deal<->contact, task<->deal)
 * - Contact props: firstname, lastname, company, jobtitle (role), email, hs_linkedin_url
 * - Deal props: dealname (title), description (comments), pipeline+dealstage (defaults for v1)
 * - Task props: hs_task_subject, hs_task_body, hs_timestamp (ms epoch UTC), hs_task_status
 * - Associations use HUBSPOT_DEFINED + numeric typeId (3=deal->contact, 216=task->deal)
 * - hs_timestamp: milliseconds since epoch (we use 09:00 UTC on due date to parallel Bitrix)
 *
 * Error handling and {id: string} returns match Bitrix24Client style exactly.
 * IDs from HubSpot are strings.
 *
 * Constructor takes access token (string). (Env/config selection is T003 scope.)
 * No new runtime dependencies.
 */

import type { CrmClient, CrmContact, CrmDeal, CrmTask } from './types.js';

export class HubspotClient implements CrmClient {
  private readonly baseUrl = 'https://api.hubapi.com';

  constructor(private readonly accessToken: string) {
    if (!accessToken || typeof accessToken !== 'string') {
      throw new Error('HubSpot access token is required');
    }
  }

  // ─── Core HTTP ───────────────────────────────────────────────────────────────

  private async call<T>(path: string, body: unknown): Promise<T> {
    const url = `${this.baseUrl}${path}`;

    let response: Response;
    try {
      response = await fetch(url, {
        method: 'POST',
        headers: {
          'Authorization': `Bearer ${this.accessToken}`,
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(body),
      });
    } catch (err) {
      throw new Error(`Network error calling HubSpot (${path}): ${String(err)}`);
    }

    let bodyJson: any;
    try {
      bodyJson = await response.json();
    } catch {
      bodyJson = {};
    }

    if (!response.ok) {
      const msg = bodyJson.message ?? bodyJson.error ?? response.statusText;
      const corr = bodyJson.correlationId ? ` (corr: ${bodyJson.correlationId})` : '';
      throw new Error(
        `HubSpot HTTP ${response.status} on ${path}: ${msg}${corr}`
      );
    }

    // HubSpot v3 error responses can have status: 'error' even on 200 in some cases
    if (bodyJson && bodyJson.status === 'error') {
      throw new Error(
        `HubSpot API error on ${path}: ${bodyJson.message ?? 'unknown error'}`
      );
    }

    if (bodyJson.id === undefined) {
      // For create, id should be present at top level
      throw new Error(`HubSpot returned no id for ${path}`);
    }

    return bodyJson as T;
  }

  // ─── Helpers ─────────────────────────────────────────────────────────────────

  // Split name like Bitrix24Client (first + rest as last). Defensive.
  private splitName(full: string): { firstname: string; lastname: string } {
    const parts = full.trim().split(/\s+/);
    return {
      firstname: parts[0] ?? full,
      lastname: parts.slice(1).join(' '),
    };
  }

  // Convert YYYY-MM-DD or ISO dueDate to ms-since-epoch UTC.
  // Uses 09:00:00Z to keep consistent semantics with Bitrix DEADLINE T09:00.
  // Per contracts + ARCHITECTURE: hs_timestamp = due ms epoch.
  private toHubspotTimestamp(dueDate: string): number {
    const base = dueDate.includes('T') ? dueDate.split('T')[0] : dueDate;
    // Construct at 09:00 UTC
    const iso = `${base}T09:00:00.000Z`;
    const ts = Date.parse(iso);
    if (isNaN(ts)) {
      // Fallback: treat as already epoch or current (should not happen in normal flow)
      return Date.now();
    }
    return ts;
  }

  // ─── Contact ─────────────────────────────────────────────────────────────────

  async createContact(contact: CrmContact): Promise<{ id: string }> {
    const { firstname, lastname } = this.splitName(contact.name);

    const properties: Record<string, string> = {
      firstname,
      lastname,
      company: contact.company,
    };
    if (contact.role) {
      properties.jobtitle = contact.role;
    }
    if (contact.email) {
      properties.email = contact.email;
    }
    if (contact.linkedin) {
      // hs_linkedin_url is the native HubSpot contact property for LinkedIn profile
      properties.hs_linkedin_url = contact.linkedin;
    }

    // POST /crm/v3/objects/contacts
    // No associations needed at create for contact (deal will link back)
    const result = await this.call<{ id: string }>('/crm/v3/objects/contacts', {
      properties,
    });

    return { id: result.id };
  }

  // ─── Deal ─────────────────────────────────────────────────────────────────────

  async createDeal(deal: CrmDeal): Promise<{ id: string }> {
    const properties: Record<string, string> = {
      dealname: deal.title,
      // description mirrors Bitrix COMMENTS field for the lead context
      description: deal.comments,
      // v1 defaults: use "default" pipeline + first common stage.
      // (Users with custom pipelines may need T003+ config; documented in comments.)
      pipeline: 'default',
      dealstage: 'appointmentscheduled',
    };

    const body = {
      properties,
      // Associate deal -> contact using the standard HUBSPOT_DEFINED type
      // TypeId 3 = Deal to contact (see associations docs + ARCHITECTURE)
      associations: [
        {
          to: { id: deal.contactId },
          types: [
            {
              associationCategory: 'HUBSPOT_DEFINED',
              associationTypeId: 3,
            },
          ],
        },
      ],
    };

    // POST /crm/v3/objects/deals
    const result = await this.call<{ id: string }>('/crm/v3/objects/deals', body);

    return { id: result.id };
  }

  // ─── Task ─────────────────────────────────────────────────────────────────────

  async createTask(task: CrmTask): Promise<{ id: string }> {
    const hsTimestamp = this.toHubspotTimestamp(task.dueDate);

    const properties: Record<string, unknown> = {
      hs_task_subject: task.title,
      hs_task_body: task.description,
      // hs_timestamp in ms since epoch UTC (required; supports number or string)
      hs_timestamp: hsTimestamp,
      hs_task_status: 'NOT_STARTED',
      // hs_task_type omitted (defaults to TODO in HubSpot)
    };

    // Associate task -> deal (216 = Task to deal)
    // We only have dealId in CrmTask (no contactId passed from orchestration).
    // Per task spec "associate to deal+contact if possible": associate to deal here.
    // (Contact association 204 would require extending CrmTask with optional contactId.)
    const associations = [
      {
        to: { id: task.dealId },
        types: [
          {
            associationCategory: 'HUBSPOT_DEFINED',
            associationTypeId: 216,
          },
        ],
      },
    ];

    // POST /crm/v3/objects/tasks
    const result = await this.call<{ id: string }>('/crm/v3/objects/tasks', {
      properties,
      associations,
    });

    return { id: result.id };
  }
}
